### PR TITLE
ARROW-6633: [C++] Vendor double-conversion library

### PR DESCRIPTION
**Pull request by pitrou**
Since this is both a mandatory dependency and a small-ish library, vendor it to make the build chain simplier. Also, make its use private, because of Windows DLL exports. This incurs a small (~8-10%) performance hit on specific workloads.

### CODE REVIEW

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
You probably want to update NOTICE.txt with the licence. Other than that, great to see so many red lines in  the CI parts.

**Comment by pitrou** [comment]
Well, it seems there's a problem with DLLs exports on Windows: https://ci.appveyor.com/project/pitrou/arrow/builds/28860108/job/ml10fmwgf9ck6bp6#L1089

File: cpp/src/arrow/util/parsing.cc [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+namespace internal {
+
+struct StringToFloatConverter::Impl {
+  Impl()
+      : main_converter_(flags_, main_junk_value_, main_junk_value_, "inf", "nan"),
+        fallback_converter_(flags_, fallback_junk_value_, fallback_junk_value_, "inf",
+                            "nan") {}
+
+  // This is only supported in double-conversion 3.1+
+#ifdef DOUBLE_CONVERSION_HAS_CASE_INSENSIBILITY
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/parsing.cc]: You don't need this anymore if you vendor it.

Comment (pitrou) [reply cpp/src/arrow/util/parsing.cc]: Hmm, true.

File: cpp/src/arrow/symbols.map [hunk 2]
```diff
@@ -23,7 +23,7 @@
       # Unfortunately this will also catch template specializations
       # (from e.g. STL or Flatbuffers) involving Arrow types.
       *arrow::*;
-      *arrow_vendored::*;
+      *arrow_vendored*;
```
Comment (fsaintjacques) [inline cpp/src/arrow/symbols.map]: Do you still need this change?

Comment (pitrou) [reply cpp/src/arrow/symbols.map]: Perhaps not :-) Thank you.

**Review by fsaintjacques — APPROVED** [review]

**Pull request closed without merging**